Implement TMD generation when packing a CIA

WorstAquaPlayer opened the pull request:
### Description

When generating a new CIA using `NodeContainer2BinaryCia`, it will now update the TMD and later convert it to binary using `TitleMetadata2Binary`. Also added class `ContentInfoRecord` to help this.
Tested generating a modified CIA for Phoenix Wright: Ace Attorney Trilogy, which installs correctly on real hardware (Old 3DS).

### Example

```cs
nodeContainer.TransformWith<NodeContainer2BinaryCia>();
```

This closes #13 and closes #14

WorstAquaPlayer committed:
:sparkles: Create a new TMD when converting a NodeContainer to a CIA
This will allow to custom content that has different size to the original one to be packed

WorstAquaPlayer committed:
:hammer: Refactor NodeContainer2BinaryCia.
It introduces classes ContentInfoRecord and TitleMetadata2Binary and makes use of them. It's still somewhat based on the previous commit code, but less spaghetti.

pleonex commented on src/Lemon/Containers/Converters/NodeContainer2BinaryCia.cs:
```diff
@@ -116,6 +131,77 @@ void WriteFile(DataWriter writer, Node file, bool isRequired = true)
+        void UpdateTitleMetadata(TitleMetadata title, Node content)
+        {
+            int chunksHashed = 0;
+            var infoRecords = new DataStream();
```
Add `using`, so the Stream is disposed/closed once the method finishes. Like `using var infoRecords = new DataStream();`

pleonex commented on src/Lemon/Containers/Converters/NodeContainer2BinaryCia.cs:
```diff
@@ -116,6 +131,77 @@ void WriteFile(DataWriter writer, Node file, bool isRequired = true)
+
+                var infoRecord = title.InfoRecords[i];
+
+                var chunksToHash = new DataStream();
```
Add `using` here too, to dispose the stream once we don't need it anymore.

pleonex commented on src/Lemon/Containers/Converters/NodeContainer2BinaryCia.cs:
```diff
@@ -77,11 +78,25 @@ public BinaryFormat Convert(NodeContainerFormat source)
+            binaryTitle.Stream.WriteTo(writer.Stream);
+            writer.Endianness = EndiannessMode.LittleEndian;
+
+            writer.Write(new byte[0x1c]);
```
I think this is needed because the padding it's missing. You need to add `writer.WritePadding(0x00, BlockSize);` like we have in `WriteFile`.

pleonex commented on src/Lemon/Containers/Converters/BinaryCia2NodeContainer.cs:
```diff
@@ -75,6 +75,8 @@ public NodeContainerFormat Convert(BinaryFormat source)
             long contentOffset = (tmdOffset + header.TitleMetaLength).Pad(BlockSize);
             Node content = UnpackContent(title, stream, contentOffset);
 
+            title = title.TransformWith<Binary2TitleMetadata>();
```
I prefer if we don't do this conversion here and we have every node by default in binary format. Also it's causing some tests to fail because the tests expects binary formats after a CIA. (See comment below for alternative).

pleonex commented on src/Lemon/Containers/Converters/NodeContainer2BinaryCia.cs:
```diff
@@ -77,11 +78,25 @@ public BinaryFormat Convert(NodeContainerFormat source)
 
             Node root = source.Root;
+
+            var title = root.Children["title"].GetFormatAs<TitleMetadata>();
```
In case we modified the `title` node from a file, the format may not be always `TitleMetadata` but `BinaryFormat`. I suggest we do a check here and convert if needed. Something like:
```csharp
if (root.Children["title"].Format is IBinary) {
    root.Children["title"].TransformWith<Binary2TitleMetadata>();
}

if (root.Children["title"].Format is not TitleMetadata) {
    throw new FormatException("Unknown format for title node");
}

var title = root.Children["title"].GetFormatAs<TitleMetadata>();
```

With this change, we don't need to convert in `Binary2Cia` anymore.

pleonex reviewed (changes requested):
Thanks a lot! The implementation looks great. Just a couple of comments regarding some minor improvements.

WorstAquaPlayer committed:
Apply comment corrections.
Added missing using statements and "title" doesn't get converted in BinaryCia2NodeContainer anymore. Also, small change in the TitleMetadata2Binary summaries.

pleonex reviewed (approved):
All good. Thanks!
